### ISSUE TITLE
Generalized svd throws

### ISSUE THREAD

**Issue by andreasnoack** [JuliaLang/julia#29553]
Refiling https://github.com/JuliaMatrices/LAPACK.jl/issues/6 here
```julia
julia> A = [-0.33872753963694624 1.124096715384297 -0.6293570718176809; 0.03919190688122216 -0.1300617417823436 0.07281871376668783];

julia> B = [-1.5303758632785613 5.136068273894432 -2.9372584484394606; 0.5364872797265587 -2.4543618264129545 2.0986693466314685];

julia> U,V,Q,C,S,R = svd(A,B)
ERROR: LAPACKException(1)
Stacktrace:
 [1] chklapackerror at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.0/LinearAlgebra/src/lapack.jl:38 [inlined]
 [2] ggsvd3!(::Char, ::Char, ::Char, ::Array{Float64,2}, ::Array{Float64,2}) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.0/LinearAlgebra/src/lapack.jl:1868
 [3] svd!(::Array{Float64,2}, ::Array{Float64,2}) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.0/LinearAlgebra/src/svd.jl:277
 [4] svd(::Array{Float64,2}, ::Array{Float64,2}) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.0/LinearAlgebra/src/svd.jl:281
 [5] top-level scope at none:0
```
I think this might be an issue with the LAPACK routine and not our wrappers. The error means

> the Jacobi-type procedure failed to converge.  For further details, see subroutine DTGSJA.

and the error is DTGSJA means

> the procedure does not converge after MAXIT cycles.

cc: @alanedelman

**Comment by ianzwaan** [comment]
Both of you are almost certainly correct. What I suspect happens is the following.

According to the documentation of xGGSVD3 `size(C, 2)` should equal `rank([A; B])`, which would be `min(m1+m2,r) == 2` in exact arithmetic.  However, in the preprocessing step (xGGSVP3) `B` is first changed to `B = V * [0 B12; 0 0] * Z'` using QR with pivoting, a rank decision, and an RQ decomposition. Then `A` is updated and partitioned such that `[A11 A12] = A * Z'`, and the whole QR+pivoting, rank decision, and RQ thing is repeated for `A11`.  This roughly results in `rank([A; B])` being computed as `rank(A11) + rank(B12)`, which may be `3` because QR with pivoting isn't as good for rank decisions as the SVD and because rank decisions are ill-posed anyway.

The problem with xTGSJA also has to do with floating-point woes, but is unrelated otherwise. For context, xTGSJA works on n-by-n upper-triangular matrices `A` and `B`, where `B` is additionally nonsingular, and iterates over a bunch of 2x2 subproblems. The routine stops iterating when the rows of `A` and `B` are approximately parallel (so that `A * inv(B)` is approximately diagonal). The tolerance used here is quite small, something like `2 * eps() * min(norm(A), norm(B))` for the 2x2 problem from this issue. This is a problem because xTGSJA uses xLAGS2 to solve the 2x2 subproblems, but the rows of the output of xLAGS2 aren't guaranteed to be THAT close to being parallel; see [Bai and Demmel '93](http://www.netlib.org/lapack/lawnspdf/lawn46.pdf) for details.

TL;DR: these are LAPACK issues, not Julia issues.

**Comment by ViralBShah** [comment]
Should we file this upstream?

**Comment by ianzwaan** [comment]
I'm not sure. Rank decisions are ill-posed problems and can be notoriously difficult, especially, in my experience, when rank decisions of submatrices are involved that depend on other rank decisions. The second problem could be addressed with a more relaxed tolerance; however, then you'd have to decide what convergence means in the face of round-off errors and if it is acceptable to stop iterating when the columns of `A` and `B` are not within machine precision of being parallel.

It could be interesting to talk to the LAPACK people about this, but at the same time I think it's important to realize that methods like this can fail (like, for example, partial pivoting in the LU decomposition).

**Comment by giordano** [comment]
Is this still relevant?
```julia-repl
julia> using LinearAlgebra

julia> A = [-0.33872753963694624 1.124096715384297 -0.6293570718176809; 0.03919190688122216 -0.1300617417823436 0.07281871376668783];

julia> B = [-1.5303758632785613 5.136068273894432 -2.9372584484394606; 0.5364872797265587 -2.4543618264129545 2.0986693466314685];

julia> U,V,Q,C,S,R = svd(A,B)
GeneralizedSVD{Float64, Matrix{Float64}, Float64, Vector{Float64}}
U factor:
2×2 Matrix{Float64}:
  0.993373  0.114937
 -0.114937  0.993373
V factor:
2×2 Matrix{Float64}:
 0.996395    0.0848337
 0.0848337  -0.996395
Q factor:
3×3 Matrix{Float64}:
 0.880933  -0.439222   0.176186
 0.403708   0.503213  -0.764066
 0.246936   0.744218   0.620614
D1 factor:
2×2 Matrix{Float64}:
 0.224609  0.0
 0.0       7.8271e-17
D2 factor:
2×2 Matrix{Float64}:
 0.974449  0.0
 0.0       1.0
R0 factor:
2×3 Matrix{Float64}:
 0.0  1.1028  -5.86746
 0.0  0.0     -3.77091

julia> versioninfo()
Julia Version 1.12.0-DEV.1054
Commit 5230d27de9* (2024-08-16 07:33 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin23.4.0)
  CPU: 8 × Apple M1
  WORD_SIZE: 64
  LLVM: libLLVM-18.1.7 (ORCJIT, apple-m1)
Threads: 1 default, 0 interactive, 1 GC (on 4 virtual cores)
```
